README.md
# Proyecto de residencias
Este es un proyecto realizado en React con NextJS y TypeScript
## Requisitos
- Tener instalado git.
- Tener instalado MySQL Workbench con MySQL Server.
> En la instalación de MySQL Workbench, al momento de llegar a la configuración del MySQL Server, en el apartado de métodos de autenticación, seleccione la opción "Use Legacy Authentication Method", puesto que el módulo MySQL para React no soporta aún el nuevo método de autenticación.
- Tener instalado Node.js
- Tener instalado Visual Studio Code o algún otro editor de código.
## Cómo ejecutar este proyecto
> Aclaración, antes de realizar todos estos pasos, deberá crear un schema en MySQL Workbench y dentro del schema crear una tabla llamada "users" y otra llamada "products". Recomiendo nombrar al schema como "proyecto-de-residencias".
1. Primero, ubique el directorio al cual desee clonar el repositorio
2. Luego, abra la terminal y clone el proyecto con
```bash
$ git clone https://github.com/rickruad/proyecto-de-residencias
```
3. Abra el directorio en donde está clonado el proyecto
```bash
$ cd ./proyecto-de-residencias/
```
4. En este directorio vamos a instalar los módulos del servidor, ejecute el siguiente comando
```bash
$ npm install
```
5. Luego, diríjase al directorio "client"
```bash
$ cd ./client/
```
6. Ahora vamos a instalar los módulos del cliente con el mismo comando
```bash
$ npm install
```
7. En MySQL Workbench, modifique las tablas "users" y "products" creadas previamente y añada las siguientes columnas

- Users

| Column Name    | Datatype     | PK  | NN  | UQ  | B   | UN  | ZF  | AI  | G   | Default/Expression |
| -------------- | ------------ | --- | --- | --- | --- | --- | --- | --- | --- | ------------------ |
| id             | INT          | Yes | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| email          | VARCHAR(45)  | No  | Yes | Yes | No  | No  | No  | No  | No  | Empty              |
| password       | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| username       | VARCHAR(45)  | No  | Yes | Yes | No  | No  | No  | No  | No  | Empty              |
| birthdate      | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| profilePicture | VARCHAR(256) | No  | No  | No  | No  | No  | No  | No  | No  | Empty                   |
| status         | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | '0'                |
| admin          | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | '0'                |

- Products

| Column Name | Datatype     | PK  | NN  | UQ  | B   | UN  | ZF  | AI  | G   | Default/Expression |
| ----------- | ------------ | --- | --- | --- | --- | --- | --- | --- | --- | ------------------ |
| id          | VARCHAR(100) | Yes | Yes | Yes | No  | No  | No  | No  | No  | Empty              |
| product     | VARCHAR(256) | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| image       | VARCHAR(256) | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| description | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| price       | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | '0'                |
| category    | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              |
| type        | VARCHAR(45)  | No  | Yes | No  | No  | No  | No  | No  | No  | Empty              | 

> Para ver la tabla, seleccione el schema creado (según lo recomendado, debería llamarse proyecto-de-residencias), cree un nuevo SQL File y escriba "SELECT * FROM users", luego le da al botón "Execute the statement under the keyboard cursor" y con eso debería de mostrarse la tabla

> En Products, la columna "type" tendría los tipos "gift-card" y "product" por ahora.
8. Ahora abra con Visual Studio Code el proyecto y diríjase al archivo index.js ubicado en ./proyecto-de-residencias/server/index.js
9. Dentro del archivo, revise las siguientes líneas
```js
const db = mysql.createConnection({
  user: 'root',
  host: 'localhost',
  password: '12345asd',
  database: 'proyecto-de-residencias',
  PORT: 3306
})
```
10. Reemplace la password con su password de su base de datos, al igual que el nombre de la database con el nombre del schema creado anteriormente (si uso el nombre recomendado, no hace falta cambiarlo).
> El PORT no es necesario cambiarlo a no ser que usted haya cambiado el puerto en donde se ejecuta MySQL Server, si ese fue el caso, reemplace el número del PORT por el que ingresó usted al momento de configurar el MySQL Server.
11. Por último, volviendo a la terminal, regrese a la raíz del proyecto (siendo este ./proyecto-de-residencias/) e inicie el servidor usando el siguiente comando
```bash
npm start
```
12. Y también inicie la webapp moviéndose al directorio ./client/ y ejecutando el siguiente comando
```bash
npm run dev
```
> El servidor se ejecutará en http://localhost:3001/ y la webapp en http://localhost:3000/

CHANGELOG.md
# Changelog
## 04-04-2023
- Añadido el soporte de fotos de perfil.
- Añadida la capacidad de agregar productos.
- La separación de categorías ahora está funcionando.
- Arreglos menores.
### Errores conocidos
- **(roto)** No se pueden editar los valores de los usuarios desde la web
- **(arreglado)** En la página de entretenimiento, en la parte de la selección de precios de las gift cards, se van a mostrar los precios del primer producto nomás. O sea que si hay 3 productos, los 3 van a tener el mismo precio del primero.

## 30-03-2023
- Arreglado el login, ahora se puede iniciar sesión y si la página detecta que entraste a las páginas de registro o login habiendo iniciado sesión, te regresa al home (funciona en viceversa).
- Agregado nuevos carteles de errores en la página del login y del registro.
- Agregada una nueva página de información de cuenta, esta cuenta con la información del usuario logueado y la posibilidad de cambiar su información. Si la cuenta es admin, aparecerá una nueva sección para administrar cuentas, en donde podrá promover, eliminar y editar cuentas.
- Agregada 2 secciones nuevas en el home, productos y servicios.
- Agregada una nueva página de categorías, por ahora solo contiene 5 categorías.
- Agregada la página de entretenimiento, no completa, solo saca datos de la base de datos, nada más.
### Errores conocidos
- Por ahora no hay cartel de error al momento de iniciar sesión con datos repetidos, tampoco al momento de editar los datos e ingresar datos repetidos.
- En la página de entretenimiento, en la parte de la selección de precios de las gift cards, se van a mostrar los precios del primer producto nomás. O sea que si hay 3 productos, los 3 van a tener el mismo precio del primero.

server/index.js
const PORT = process.env.PORT || 3001;

const path = require('path');
const cors = require('cors');
const mysql = require('mysql');
const multer = require('multer');
const express = require('express');

const app = express();

app.use(cors());
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

const db = mysql.createConnection({
  user: 'root',
  host: 'localhost',
  password: '12345asd',
  database: 'proyecto-de-residencias',
  PORT: 3306
})

db.connect();

const storageAvatars = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, "./client/public/img/profile-pictures/");
  },
  filename: (req, file, cb) => {
    const originalName = file.originalname;
    const extension = originalName.split(".").pop();
    const timestamp = new Date().toISOString().replace(/[-T:\.Z]/g, "");
    const fileName = `${originalName.split(".")[0]}-${timestamp}.${extension}`;
    cb(null, fileName);
  },
});

const uploadAvatars = multer({ storage: storageAvatars });

app.post('/api/sing-up', uploadAvatars.single("image"), (req, res) => {
  const username = req.body.username;
  const birthdate = req.body.birthdate;
  const email = req.body.email;
  const password = req.body.password;

  const imageUrl = `/img/profile-pictures/${req.file.filename}`;
  const query = 'INSERT INTO users (id, username, birthdate, email, password, profilePicture) VALUES (?,?,?,?,?,?)'

  const insertData = (idToTry) => {
    db.query(query, [idToTry, username, birthdate, email, password, imageUrl], (err, result) => {
      if (err) {
        if (err.code === 'ER_DUP_ENTRY') {
          insertData(parseInt(idToTry) + 1);
        } else {
          console.log(err);
        }
      } else {
        res.send({ message: 'Cuenta registrada con éxito' });
      }
    });
  }

  insertData(1);
});

app.post('/api/sing-in', (req, res) => {
  const email = req.body.email;
  const password = req.body.password;

  db.query("SELECT * FROM users WHERE email = ? AND password = ?", [email, password], (err, result) => {
    if (err) {
      console.log(err);
    } else {
      if (result.length > 0) {
        db.query("UPDATE users SET status = '1' WHERE email = ? AND password = ?", [email, password], (err, updateResult) => {
          if (err) {
            console.log(err);
          } else {
            console.log(updateResult);
          }
        });
        res.send(result);
        console.log(result);
      } else {
        res.send({ message: "Email o contraseña incorrecta" });
        console.log('Email o contraseña incorrecta');
      }
    }
  });
})

app.post('/api/sing-out', (req, res) => {
  db.query("SELECT * FROM users WHERE status = '1'", (err, result) => {
    if (err) {
      console.log(err);
    } else {
      if (result.length > 0) {
        db.query("UPDATE users SET status = '0' WHERE status = '1'", (err, updateResult) => {
          if (err) {
            console.log(err);
          } else {
            console.log(updateResult);
            res.send({ message: 'Success' });
          }
        });
      } else {
        console.log('error inesperado');
        console.log(result)
      }
    }
  })
})

app.post('/api/user', (req, res) => {  
  db.query("SELECT * FROM users WHERE status = '1'", (err, result) => {
    if (err) {
      console.log(err);
    } else {
      if (result.length > 0) {
        res.send(result);
      } else {
        res.json({ message: 'USER ERROR' })
      }
    }
  });
});

app.post('/api/users', (req, res) => {
  db.query("SELECT * FROM users", (err, result) => {
    if (err) {
      console.log(err);
    } else {
      if (result.length > 0) {
        res.send(result);
      } else {
        res.json({ message: 'USER ERROR' })
      }
    }
  })
})

app.post('/api/user-status', (req, res) => {
  db.query("SELECT * FROM users WHERE status = '1'", (err, result) => {
    if (err) {
      console.log(err);
    } else {
      if (result.length > 0) {
        res.send({ message: 1 });
      } else {
        res.send({ message: 0 });
      }
    }
  });
});

app.post('/api/delete-user', (req, res) => {
  const username = req.body.username;

  db.query("DELETE FROM users WHERE username = ?", [username], (err, result) => {
    if (err) {
      console.log(err);
    }
  })
})

app.post('/api/promote-user', (req, res) => {
  const username = req.body.username;

  db.query("UPDATE users SET admin = '1' WHERE username = ? ", [username], (err, result) => {
    if (err) {
      console.log(err);
    }
  })
})

app.post('/api/update-user', uploadAvatars.single("image"), (req, res) => {
  const id = req.body.id;
  const email = req.body.email;
  const password = req.body.password;
  const username = req.body.username;
  const birthdate = req.body.birthdate;
  const oldProfilePicture = req.body.profilePicture;
  const profilePicture = `/img/profile-pictures/${req.file.filename}`;

  const query = 'UPDATE users SET email = ?, password = ?, username = ?, birthdate = ?, profilePicture = ? WHERE id = ?';
  const values = [email, password, username, birthdate, profilePicture ? profilePicture : oldProfilePicture, id];

  db.query(query, values, (err, result) => {
    if (err) {
      console.log(err);
      res.send({ message: sqlMessage })
    } else {
      res.send({ message: 'Actualizado los valores correctamente' });
    }
  })
})

app.post("/api/server-status", (req, res) => {
  res.json({ message: "El servidor está en linea" });
});

app.post("/api/get-products", (req, res) => {
  db.query("SELECT * FROM products", (err, result) => {
    if(err) {
      console.log(err);
    } else {
      if(result.length > 0) {
        res.send(result);
      } else {
        res.json({ message: 'PRODUCT ERROR' });
      }
    }
  })
})

const storageProduct = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, "./client/public/img/products-pictures/");
  },
  filename: (req, file, cb) => {
    const originalName = file.originalname;
    const extension = originalName.split(".").pop();
    const timestamp = new Date().toISOString().replace(/[-T:\.Z]/g, "");
    const fileName = `${originalName.split(".")[0]}-${timestamp}.${extension}`;
    cb(null, fileName);
  },
});

const uploadProduct = multer({ storage: storageProduct });

app.post("/api/add-product", uploadProduct.single("image"), (req, res) => {
  const name = req.body.productName;
  const description = req.body.productDescription;
  const price = req.body.productPrice;
  const category = req.body.productCategory;
  const type = req.body.productType;

  const imageUrl = `/img/products-pictures/${req.file.filename}`;

  const query = 'INSERT INTO products (id, product, image, description, price, category, type) VALUE (?, ?, ?, ?, ?, ?, ?)';

  const insertData = (idToTry) => {
    db.query(query, [idToTry, name, imageUrl, description, price, category, type], (err, result) => {
      if (err) {
        if (err.code === 'ER_DUP_ENTRY') {
          insertData(parseInt(idToTry) + 1);
        } else {
          console.log(err);
        }
      } else {
        res.send({ message: 'Producto agregado con exito' });
      }
    })
  }

  insertData(1);
})

app.listen(PORT, () => {
  console.log(`server running on http://localhost:${PORT}/`);
});